url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+        }
+      }
+
+      addToSet(NO_SPEC_INT, -1, 0, type);
+      return i;
+    }
+
+    if (c == '*' || c == '/') {
+      if (c == '*' && (i + 1) >= s.length()) {
+        addToSet(ALL_SPEC_INT, -1, incr, type);
+        return i + 1;
+      } else if (c == '/' && ((i + 1) >= s.length() || s.charAt(i + 1) == ' ' || s.charAt(i + 1) == '\t')) {
+        throw new ParseException("'/' must be followed by an integer.", i);
+      } else if (c == '*') {
+        i++;
+      }
+      c = s.charAt(i);
+      if (c == '/') { // is an increment specified?
+        i++;
+        if (i >= s.length()) {
+          throw new ParseException("Unexpected end of string.", i);
+        }
+
+        incr = getNumericValue(s, i);
+
+        i++;
+        if (incr > 10) {
+          i++;
+        }
+        if (incr > 59 && (type == SECOND || type == MINUTE)) {
+          throw new ParseException("Increment > 60 : " + incr, i);
+        } else if (incr > 23 && (type == HOUR)) {
+          throw new ParseException("Increment > 24 : " + incr, i);
+        } else if (incr > 31 && (type == DAY_OF_MONTH)) {
+          throw new ParseException("Increment > 31 : " + incr, i);
+        } else if (incr > 7 && (type == DAY_OF_WEEK)) {
+          throw new ParseException("Increment > 7 : " + incr, i);
+        } else if (incr > 12 && (type == MONTH)) {
+          throw new ParseException("Increment > 12 : " + incr, i);
+        }
+      } else {
+        incr = 1;
+      }
+
+      addToSet(ALL_SPEC_INT, -1, incr, type);
+      return i;
+    } else if (c == 'L') {
+      i++;
+      if (type == DAY_OF_MONTH) {
+        lastdayOfMonth = true;
+      }
+      if (type == DAY_OF_WEEK) {
+        addToSet(7, 7, 0, type);
+      }
+      if (type == DAY_OF_MONTH && s.length() > i) {
+        c = s.charAt(i);
+        if (c == '-') {
+          ValueSet vs = getValue(0, s, i + 1);
+          lastdayOffset = vs.value;
+          if (lastdayOffset > 30) {
+            throw new ParseException("Offset from last day must be <= 30", i + 1);
+          }
+          i = vs.pos;
+        }
+        if (s.length() > i) {
+          c = s.charAt(i);
+          if (c == 'W') {
+            nearestWeekday = true;
+            i++;
+          }
+        }
+      }
+      return i;
+    } else if (c >= '0' && c <= '9') {
+      int val = Integer.parseInt(String.valueOf(c));
+      i++;
+      if (i >= s.length()) {
+        addToSet(val, -1, -1, type);
+      } else {
+        c = s.charAt(i);
+        if (c >= '0' && c <= '9') {
+          ValueSet vs = getValue(val, s, i);
+          val = vs.value;
+          i = vs.pos;
+        }
+        i = checkNext(i, s, val, type);
+        return i;
+      }
+    } else {
+      throw new ParseException("Unexpected character: " + c, i);
+    }
+
+    return i;
+  }
+
+  protected int checkNext(int pos, String s, int val, int type) throws ParseException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 652-700, lines 702-742, lines 744-763 into separate methods.